extensions/amp-consent/0.1/amp-consent.js

```diff
@@ -723,6 +729,102 @@ export class AmpConsent extends AMP.BaseElement {
       opt_metadata['gdprApplies']
     );
   }
+
+  /**
+   * Checks if the payload from the `tcfapiCall` is valid.
+   * @param {JsonObject} payload
+   * @return {boolean}
+   */
+  isValidTcfApiCall_(payload) {
+    if (!isObject(payload)) {
+      user().error(TAG, `"tcfapiCall" is not an object: ${payload}`);
+      return false;
+    }
+    const {command, parameter, version} = payload;
+    if (!isEnumValue(TCF_POST_MESSAGE_API_COMMANDS, command)) {
+      user().error(
+        TAG,
+        `Unsupported command found in "tcfapiCall": ${command}`
+      );
+      return false;
+    }
+    if (parameter !== undefined) {
+      user().error(
+        TAG,
+        `Unsupported parameter found in "tcfapiCall": ${parameter}`
+      );
+      return false;
+    }
+    if (version != '2') {
+      user().error(TAG, `Found incorrect version in "tcfapiCall": ${version}`);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Maybe set up the __tfcApiLocator window and listeners.
+   *
+   * The window is a dummy iframe that signals to 3p iframes
+   * that the document supports the tcfPostMessage API.
+   */
+  maybeSetUpTcfPostMessageProxy_() {
+    if (!this.consentConfig_['exposesTcfApi']) {
+      return;
+    }
+    // Check if __tcfApiLocator API already exists (dirty AMP)
+    if (!this.win.frames[TCF_API_LOCATOR]) {
+      // Add window listener for 3p iframe PostMessages
+      this.win.addEventListener('message', this.boundHandleIframeMessages_);
```

Comment:
In theory, it's still possible that the message arrives before the event listener is registered. Or the 3p code start to looking for `TCF_API_LOCATOR` before the iframe is created. What does the TCF API suggest in this scenario?

Reply:
When is the 3p setup?

Reply:
> In theory, it's still possible that the message arrives before the event listener is registered. Or the 3p code start to looking for TCF_API_LOCATOR before the iframe is created. What does the TCF API suggest in this scenario?

There is no recommendation for what happens if the parent doc event listener is not registered before the message arrives. 

If the `TCF_API_LOCATOR` window is not present when the iframe loads, it is assumed that no CMP is present (and so the tcfApi is not supported). 

In theory, both of these could happen but only if the publisher does not tag the element with `data-block-on-consent`. For this API we state that elements that create iframes that could use this API need to be blocked with `data-block-on-consent` (as they already should be if they want to collect consent information anyway). Additionally, CMPs working with publishers should tell them to include this `exposesTcfApi` and also block certain elements with `data-block-on-consent`.

> When is the 3p setup?

3p setup occurs once the other extensions (amp-iframe, amp-video-iframe etc) loads the iframe. We state that for this API to work, these components need to be blocked by `data-block-on-consent` (to ensure that the `TCF_API_LOCATOR ` window is already created).